File: test_scripts/ancestral_dense.py

from Bio import AlignIO
import numpy as np
from lib import Graph, graph_from_nwk_str, GraphNodeBackward, GraphNodeForward, Node
from lib import RangeCollection_intersection, RangeCollection, Mut, SeqPartition, DenseSeqNode, DenseSeqEdge, DenseSeqInfo, DenseSeqDis, find_char_ranges
from treetime import GTR
from treetime.seq_utils import seq2prof
from profile_map import profile_map
from typing import List, Dict
from payload import NodePayload, EdgePayload

profile_map['-'] = profile_map['N']

def attach_seqs_to_graph_dense(graph: Graph, aln_list: List[Dict[str,np.array]], gtr_list: List[GTR]) -> None:
  name_to_key = {}
  for n in graph.get_leaves():
    n.payload().dense_sequences = []
    name_to_key[n.payload().name] = n.key()
  for edge in graph.get_edges():
    edge.payload().dense_sequences = []

  graph.dense_partitions = []
  for aln, gtr in zip(aln_list, gtr_list):
    L = len(aln[list(aln.keys())[0]]) # stupid way to get alignment length
    partition = SeqPartition(gtr=gtr, length=L, profile_map=profile_map)
    graph.dense_partitions.append(partition)

    for leaf in graph.get_leaves():
      leaf_name = leaf.payload().name
      seq_array = np.fromiter(aln[leaf_name], 'U1')
      seq_node = DenseSeqNode(seq=DenseSeqInfo(sequence=seq_array, gaps=find_char_ranges(seq_array, '-')),
                              msg_to_parents=DenseSeqDis(dis=seq2prof(seq_array, profile_map=profile_map)))
      leaf.payload().dense_sequences.append(seq_node)

    for edge in graph.get_edges():
      edge.payload().dense_sequences.append(DenseSeqEdge())

def init_sequences_dense(graph: Graph, aln_list: List[Dict[str,np.array]], gtr_list: List[GTR]) -> None:
  attach_seqs_to_graph_dense(graph, aln_list, gtr_list)

def combine_dense_messages(msgs):
  seq_dis = DenseSeqDis(logLH=np.sum([m.logLH for m in msgs.values()]))
  prod_dis = np.prod([m.dis for m in msgs.values()], axis=0)
  norm = prod_dis.sum(axis=1)
  seq_dis.logLH += np.sum(np.log(norm))
  seq_dis.dis = (prod_dis.T/norm).T
  return seq_dis

def ingroup_profiles_dense(graph: Graph):
  alphabets = [''.join(p.gtr.alphabet) for p in graph.dense_partitions]
  gtrs = [p.gtr for p in graph.dense_partitions]
  seq_len = [p.length for p in graph.dense_partitions]

  def calculate_ingroup_node(node: GraphNodeBackward) -> None:
    if node.is_leaf:
      return # the msg to parents for leaves was put in init phase

    node.payload.dense_sequences=[]
    for si,gtr in enumerate(gtrs):
      child_expQt = [gtr.expQt(e.branch_length or 0.0) for c,e in node.children]
      child_seqs =  [c.dense_sequences[si] for c,e in node.children]
      child_edges = [e.dense_sequences[si] for c,e in node.children]
      gap_intersection = RangeCollection_intersection([cseq.seq.gaps for cseq in child_seqs])
      seq_info = DenseSeqInfo(gaps=gap_intersection)

      msgs_from_children = {}
      for ci, (c,e) in enumerate(node.children):
        seq_dis = DenseSeqDis(dis = np.ones((seq_len[si], len(alphabets[si]))))
        seq_dis.logLH += child_seqs[ci].msg_to_parents.logLH
        if child_edges[ci].transmission:
          for r in child_edges[ci].transmission:
            seq_dis.dis[r.start:r.end] *= child_seqs[ci].msg_to_parents.dis[r.start:r.end].dot(child_expQt[ci])
        else:
          seq_dis.dis *= child_seqs[ci].msg_to_parents.dis.dot(child_expQt[ci])
        msgs_from_children[c.name] = seq_dis

      msg_to_parents = combine_dense_messages(msgs_from_children)
      node.payload.dense_sequences.append(DenseSeqNode(seq=seq_info, msgs_from_children=msgs_from_children,
                                                       msg_to_parents=msg_to_parents))

  graph.par_iter_backward(calculate_ingroup_node)

def outgroup_profiles_dense(graph: Graph):
  alphabets = [''.join(p.gtr.alphabet) for p in graph.dense_partitions]
  gtrs = [p.gtr for p in graph.dense_partitions]
  seq_len = [p.length for p in graph.dense_partitions]

  def calculate_outgroup_node(node: GraphNodeBackward) -> None:
    if node.is_root:
      return
    for si,seq_info in enumerate(node.payload.dense_sequences):
      msgs_from_parents = {}
      for p,e in node.parents:
        eQt = gtrs[si].expQt(e.branch_length or 0.0).T
        pseq = p.dense_sequences[si]
        seq_dis = DenseSeqDis(dis = np.ones((seq_len[si], len(alphabets[si]))),
                              logLH=pseq.msgs_to_children[node.payload.name].logLH)
        if e.dense_sequences[si].transmission:
          for r in e.dense_sequences[si].transmission:
            seq_dis.dis[r.start:r.end] *= pseq.msgs_to_children[node.payload.name].dis[r.start:r.end].dot(eQt)
        else:
          # could make this a copy
          seq_dis.dis *= pseq.msgs_to_children[node.payload.name].dis.dot(eQt)
        msgs_from_parents[p.name] = seq_dis
      msgs_from_parents['children'] = seq_info.msg_to_parents
      seq_info.profile = combine_dense_messages(msgs_from_parents)

      seq_info.msgs_to_children = {}
      for cname in seq_info.msgs_from_children:
        seq_info.msgs_to_children[cname] = DenseSeqDis(dis = seq_info.profile.dis/seq_info.msgs_from_children[cname].dis,
                              logLH=seq_info.profile.logLH-seq_info.msgs_from_children[cname].logLH)


  graph.par_iter_forward(calculate_outgroup_node)

def calculate_root_state_dense(graph: Graph):
  logLH=0
  for root_node in graph.get_roots():
    for si,seq_info in enumerate(root_node.payload().dense_sequences):
      gtr = graph.dense_partitions[si].gtr
      seq_dis = DenseSeqDis(logLH=seq_info.msg_to_parents.logLH)
      seq_dis.dis = seq_info.msg_to_parents.dis*gtr.Pi
      norm = seq_dis.dis.sum(axis=1)
      seq_dis.logLH += np.sum(np.log(norm))
      seq_dis.dis = (seq_dis.dis.T/norm).T

      seq_info.profile = seq_dis
      seq_info.msgs_to_children = {}
      for cname in seq_info.msgs_from_children:
        seq_info.msgs_to_children[cname] = DenseSeqDis(dis = seq_info.profile.dis/seq_info.msgs_from_children[cname].dis,
                              logLH=seq_info.profile.logLH-seq_info.msgs_from_children[cname].logLH)
      logLH+=seq_info.profile.logLH
  return logLH

def ancestral_dense(graph: Graph) -> float:
  ingroup_profiles_dense(graph)
  logLH = calculate_root_state_dense(graph)
  outgroup_profiles_dense(graph)
  return logLH


if __name__=="__main__":
  fname_nwk = 'data/ebola/ebola.nwk'
  fname_seq = 'data/ebola/ebola_dna.fasta'
  fname_nwk = 'test_scripts/data/tree.nwk'
  fname_seq = 'test_scripts/data/sequences.fasta'
  with open(fname_nwk) as fh:
    nwkstr = fh.read()
  G = graph_from_nwk_str(nwk_string=nwkstr, node_payload_factory=NodePayload, edge_payload_factory=EdgePayload)

  aln = {seq.id: str(seq.seq).upper() for seq in AlignIO.read(fname_seq, 'fasta')}
  gtr = GTR.custom(pi=[0.2, 0.3, 0.15, 0.35], alphabet='nuc_nogap')
  init_sequences_dense(G, [aln], [gtr])

  print("LogLH", ancestral_dense(G))

File: packages/treetime/src/port/ancestral_dense.rs
use crate::alphabet::alphabet::Alphabet;
use crate::graph::breadth_first::GraphTraversalContinuation;
use crate::graph::edge::Weighted;
use crate::graph::node::Named;
use crate::port::seq_dense::{DenseGraph, DenseNode, DenseSeqDis, DenseSeqEdge, DenseSeqInfo, DenseSeqNode};
use crate::port::seq_partitions::{PartitionLikelihood, PartitionLikelihoodWithAln};
use crate::seq::range_intersection::range_intersection;
use crate::utils::ndarray::{log, product_axis};
use crate::{make_internal_report, make_report, o};
use eyre::Report;
use itertools::Itertools;
use maplit::btreemap;
use ndarray::prelude::*;
use ndarray::stack;
use ndarray_stats::QuantileExt;
use std::collections::BTreeMap;

// Turn a profile into a sequence
fn prof2seq(profile: &DenseSeqDis, alphabet: &Alphabet) -> Vec<char> {
  // iterate over profile.dis and take the argmax of each row, lookup in alphabet.char to extract the character
  let mut seq = Vec::new();
  for row in profile.dis.rows() {
    let argmax = row.argmax().unwrap();
    seq.push(alphabet.char(argmax));
  }
  seq
}

fn assign_sequence(seq_info: &DenseSeqNode, alphabet: &Alphabet) -> Vec<char> {
  let mut seq = prof2seq(&seq_info.profile, alphabet);
  for gap in &seq_info.seq.gaps {
    seq[gap.0..gap.1].fill(alphabet.gap());
  }
  seq
}

fn attach_seqs_to_graph(graph: &DenseGraph, partitions: &[PartitionLikelihoodWithAln]) -> Result<(), Report> {
  graph.get_leaves().iter().try_for_each(|leaf| -> Result<(), Report>  {
    let mut leaf = leaf.write_arc().payload().write_arc();

    let leaf_name = leaf.name.as_ref().ok_or_else(|| {
      make_report!("Expected all leaf nodes to have names, such that they can be matched to their corresponding sequences. But found a leaf node that has no name.")
    })?.to_owned();

    leaf.dense_partitions = partitions
      .iter()
      .map(|PartitionLikelihoodWithAln { gtr, alphabet, aln, length }| {
        // TODO(perf): this might be slow if there are many sequences
        let leaf_fasta = aln
          .iter()
          .find(|fasta| fasta.seq_name == leaf_name)
          .ok_or_else(|| make_internal_report!("Leaf sequence not found: '{leaf_name}'"))?;

        // TODO(perf): unnecessary copy of sequence data. Neither String, nor &[char] works well for us, it seems.
        // We probably want a custom class for sequences. Sequences should be instantiated in the fasta parser and
        // never need a copy like here.
        let sequence = leaf_fasta.seq.chars().collect::<Vec<_>>();

        DenseSeqNode::new(&sequence, alphabet)
      })
      .collect::<Result<_, Report>>()?;

    Ok(())
  })?;

  graph.get_edges().iter().for_each(|edge| {
    let mut edge = edge.write_arc().payload().write_arc();
    edge.dense_partitions = partitions.iter().map(|_| DenseSeqEdge::default()).collect_vec();
  });

  Ok(())
}

fn combine_dense_messages(msgs: &BTreeMap<String, DenseSeqDis>) -> Result<DenseSeqDis, Report> {
  let stacked = stack(Axis(0), &msgs.values().map(|d| d.dis.view()).collect_vec())?;
  let prod_dis = product_axis(&stacked, Axis(0));
  let norm = prod_dis.sum_axis(Axis(1));
  let dis = (&prod_dis.t() / &norm).t().to_owned();
  let log_lh = msgs.values().map(|m| m.log_lh).sum::<f64>() + log(&norm).sum();
  Ok(DenseSeqDis { dis, log_lh })
}

fn ingroup_profiles_dense(graph: &DenseGraph, partitions: &[PartitionLikelihood]) {
  let n_partitions = partitions.len();

  graph.par_iter_breadth_first_backward(|mut node| {
    if node.is_leaf {
      return GraphTraversalContinuation::Continue; // the msg to parents for leaves was put in init phase
    }

    node.payload.dense_partitions = (0..n_partitions)
      .map(|si| {
        let PartitionLikelihood { gtr, alphabet, length } = &partitions[si];

        let gaps = node
        .children
        .iter()
        .map(|(c, e)| c.read_arc().dense_partitions[si].seq.gaps.clone()) // TODO: avoid cloning
        .collect_vec();

        let gaps = range_intersection(&gaps);

        let seq = DenseSeqInfo {
          gaps,
          ..DenseSeqInfo::default()
        };

        let mut msgs_from_children = btreemap! {};

        for (child, edge) in &node.children {
          let (child, edge) = (child.read_arc(), edge.read_arc());

          let name = child
            .name()
            .expect("Encountered child node without a name")
            .as_ref()
            .to_owned();

          let exp_qt = gtr.expQt(edge.weight().unwrap_or_default());
          let child = &child.dense_partitions[si];
          let edge = &edge.dense_partitions[si];

          let mut dis = Array2::ones((*length, alphabet.n_canonical()));
          let log_lh = child.msg_to_parents.log_lh;

          // Note that these dot products are really just
          //     a_{ik} = b.dot(c) = sum(b_{ij}c{jk}, j)
          // -- we can use whatever memory layout we want.
          if let Some(transmission) = &edge.transmission {
            for r in transmission {
              dis
                .slice_mut(s![r.0..r.1, ..])
                .assign(&child.msg_to_parents.dis.slice(s![r.0..r.1, ..]).dot(&exp_qt));
            }
          } else {
            // could make this a copy
            dis *= &child.msg_to_parents.dis.dot(&exp_qt);
          }

          msgs_from_children.insert(name, DenseSeqDis { dis, log_lh });
        }

        let msg_to_parents = combine_dense_messages(&msgs_from_children).unwrap();

        DenseSeqNode {
          seq,
          msg_to_parents,
          msgs_from_children,
          ..DenseSeqNode::default()
        }
      })
      .collect_vec();

    GraphTraversalContinuation::Continue
  });
}

fn outgroup_profiles_dense(graph: &DenseGraph, dense_partitions: &[PartitionLikelihood]) {
  graph.par_iter_breadth_first_forward(|mut node| {
    let name = node
      .payload
      .name()
      .expect("Encountered node without a name")
      .as_ref()
      .to_owned();

    if node.is_root {
      return GraphTraversalContinuation::Continue;
    }

    for (si, seq_info) in node.payload.dense_partitions.iter_mut().enumerate() {
      let PartitionLikelihood { gtr, alphabet, length } = &dense_partitions[si];

      let mut msgs_from_parents = btreemap! {};
      for (parent, edge) in &node.parents {
        let parent = parent.read_arc();
        let edge = edge.read_arc();

        let exp_qt = gtr.expQt(edge.weight().unwrap_or(0.0)).t().to_owned();
        let parent = &parent.dense_partitions[si];
        let edge = &edge.dense_partitions[si];
        let mut dis = Array2::ones((*length, alphabet.n_canonical()));
        let log_lh = parent.msgs_to_children[&name].log_lh;
        if let Some(transmission) = &edge.transmission {
          for r in transmission {
            dis
              .slice_mut(s![r.0..r.1, ..])
              .assign(&(parent.msgs_to_children[&name].dis.slice(s![r.0..r.1, ..]).dot(&exp_qt)));
          }
        } else {
          dis = &dis * &parent.msgs_to_children[&name].dis.dot(&exp_qt);
        }
        msgs_from_parents.insert(name.clone(), DenseSeqDis { dis, log_lh });
      }

      msgs_from_parents.insert(o!("children"), seq_info.msg_to_parents.clone()); // HACK
      seq_info.profile = combine_dense_messages(&msgs_from_parents).unwrap();
      seq_info.seq.sequence = assign_sequence(seq_info, alphabet);
      seq_info.msgs_to_children = btreemap![];
      for cname in seq_info.msgs_from_children.keys() {
        let child_msg = &seq_info.msgs_from_children[cname];
        let mut dis = seq_info.profile.dis.clone();
        dis /= &child_msg.dis;
        let log_lh = seq_info.profile.log_lh - child_msg.log_lh;
        seq_info
          .msgs_to_children
          .insert(cname.clone(), DenseSeqDis { dis, log_lh });
      }
    }

    GraphTraversalContinuation::Continue
  });
}

fn calculate_root_state_dense(graph: &DenseGraph, partitions: &[PartitionLikelihood]) -> f64 {
  let mut total_log_lh = 0.0;
  for root_node in graph.get_roots() {
    let dense_partitions = &mut root_node.write_arc().payload().write_arc().dense_partitions;
    for (si, seq_info) in dense_partitions.iter_mut().enumerate() {
      let PartitionLikelihood { gtr, alphabet, .. } = &partitions[si];

      let mut log_lh = seq_info.msg_to_parents.log_lh;
      let mut dis = &seq_info.msg_to_parents.dis * &gtr.pi;

      let norm = dis.sum_axis(Axis(1));
      log_lh += log(&norm).sum();
      dis = (&dis.t() / &norm).t().to_owned();
      seq_info.profile = DenseSeqDis { dis, log_lh };
      seq_info.seq.sequence = assign_sequence(seq_info, alphabet);
      seq_info.msgs_to_children.clear();
      for cname in seq_info.msgs_from_children.keys() {
        // This division operation can cause 'division by 0' problems. Note that profile = prod(msgs[child], child in children) * msgs_from_parent.
        // Alternatively, msgs_to_children could be calculated as prod(msgs[sibling], sibling in children/child) * msgs_from_parent.
        // This latter redundant calculation is done in the sparse case, but division is more efficient in particular in case of polytomies.
        let child_msg = &seq_info.msgs_from_children[cname];
        let dis = &seq_info.profile.dis / &child_msg.dis;
        let log_lh = seq_info.profile.log_lh - child_msg.log_lh;
        seq_info
          .msgs_to_children
          .insert(cname.clone(), DenseSeqDis { dis, log_lh });
      }

      total_log_lh += seq_info.profile.log_lh;
    }
  }
  total_log_lh
}

pub fn run_marginal_dense(graph: &DenseGraph, partitions: Vec<PartitionLikelihoodWithAln>) -> Result<f64, Report> {
  attach_seqs_to_graph(graph, &partitions)?;

  let partitions = partitions.into_iter().map(PartitionLikelihood::from).collect_vec();

  ingroup_profiles_dense(graph, &partitions);
  let log_lh = calculate_root_state_dense(graph, &partitions);
  outgroup_profiles_dense(graph, &partitions);
  Ok(log_lh)
}

pub fn ancestral_reconstruction_marginal_dense(
  graph: &DenseGraph,
  include_leaves: bool,
  mut visitor: impl FnMut(&DenseNode, Vec<char>),
) -> Result<(), Report> {
  graph.iter_depth_first_preorder_forward(|node| {
    if !include_leaves && node.is_leaf {
      return;
    }

    let seq = node
      .payload
      .dense_partitions
      .iter()
      .flat_map(|p| p.seq.sequence.iter().copied())
      .collect();

    visitor(&node.payload, seq);
  });

  Ok(())
}

#[cfg(test)]
mod tests {
  use super::*;
  use crate::alphabet::alphabet::AlphabetName;
  use crate::gtr::get_gtr::{jc69, JC69Params};
  use crate::io::fasta::read_many_fasta_str;
  use crate::io::json::{json_write_str, JsonPretty};
  use crate::io::nwk::nwk_read_str;
  use crate::utils::string::vec_to_string;
  use eyre::Report;
  use indoc::indoc;
  use pretty_assertions::assert_eq;

  #[test]
  fn test_ancestral_reconstruction_marginal_dense() -> Result<(), Report> {
    rayon::ThreadPoolBuilder::new().num_threads(1).build_global()?;

    let aln = read_many_fasta_str(indoc! {r#"
      >root
      TCAGCCATGTATTG--
      >AB
      ACATCCCTGTA-TG--
      >A
      ACATCGCCNNA--GAC
      >B
      GCATCCCTGTA-NG--
      >CD
      CCGGCCATGTATTG--
      >C
      CCGGCGATGTRTTG--
      >D
      TCGGCCGTGTRTTG--
    "#})?;

    let expected = read_many_fasta_str(indoc! {r#"
      >root
      TCGGCGCTGTATTGAC
      >AB
      ACATCGCTGTA-TGAC
      >CD
      TCGGCGGTGTATTG--
    "#})?
    .into_iter()
    .map(|fasta| (fasta.seq_name, fasta.seq))
    .collect::<BTreeMap<_, _>>();

    let graph: DenseGraph = nwk_read_str("((A:0.1,B:0.2)AB:0.1,(C:0.2,D:0.12)CD:0.05)root:0.01;")?;

    let treat_gap_as_unknown = true;
    let alphabet = Alphabet::new(AlphabetName::Nuc, treat_gap_as_unknown)?;
    let gtr = jc69(JC69Params {
      alphabet: AlphabetName::Nuc,
      treat_gap_as_unknown: true,
      ..JC69Params::default()
    })?;
    let partitions = vec![PartitionLikelihoodWithAln::new(gtr, alphabet, aln)?];
    run_marginal_dense(&graph, partitions)?;

    let mut actual = BTreeMap::new();
    ancestral_reconstruction_marginal_dense(&graph, false, |node, seq| {
      actual.insert(node.name.clone(), vec_to_string(seq));
    })?;

    assert_eq!(
      json_write_str(&expected, JsonPretty(false))?,
      json_write_str(&actual, JsonPretty(false))?
    );

    Ok(())
  }
}
